src/consensus/pool/finality_tracker.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright (c) Anza Technology, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Tracks finality of blocks.
+//!
+//! This is used as part of [`PoolImpl`].
+//!
+//! Keeps track of:
+//! - Direct finalization of blocks,
+//! - resulting indirect finalizations of blocks, AND
+//! - resulting implicit skipping of slots
+//!
+//! It does this based on:
+//! - Notarization of blocks,
+//! - finalization of slots, AND
+//! - availability of blocks and knowledge of their parents.
+//!
+//! [`PoolImpl`]: crate::consensus::pool::PoolImpl
+
+use std::collections::BTreeMap;
+
+use crate::BlockId;
+use crate::crypto::Hash;
+use crate::types::Slot;
+
+/// Tracks finality of blocks.
+#[derive(Default)]
+pub struct FinalityTracker {
+    /// Current finalization status for each slot.
+    status: BTreeMap<Slot, FinalizationStatus>,
+    /// Maps blocks to their parents.
+    parents: BTreeMap<BlockId, BlockId>,
+    /// The highest finalized slot so far.
+    highest_finalized: Slot,
+}
+
+/// Possible states a slot can be in regarding finality.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum FinalizationStatus {
+    Notarized(Hash),
+    Finalized,
+    FinalizedAndNotarized(Hash),
+    ImplicitlyFinalized(Hash),
+    ImplicitlySkipped,
+}
+
+/// Information about newly finalized slots.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct FinalizationEvent {
+    /// Directly finalized block, if any.
+    pub(super) finalized: Option<BlockId>,
+    /// Any implicitly finalized blocks.
+    pub(super) implicitly_finalized: Vec<BlockId>,
+    /// Any implicitly skipped slots.
+    pub(super) implicitly_skipped: Vec<Slot>,
+}
+
+impl FinalityTracker {
+    /// Adds the given `parent` for the given `block`.
+    ///
+    /// Handles possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn add_parent(&mut self, block: BlockId, parent: BlockId) -> FinalizationEvent {
+        assert!(block.0 > parent.0);
+        if let Some(p) = self.parents.get(&block) {
+            assert!(p == &parent);
+            return FinalizationEvent::default();
+        }
+        self.parents.insert(block, parent);
```

Comment:
We are doing two accesses into the map here.  We could either check the return value from `insert` or use the `entry` API.

Reply:
changed in a bunch of places, see https://github.com/qkniep/alpenglow/pull/99/commits/321657f5e01190e8426b1ee9feb07bd1b96e1698